Add support for match ... case

prsabahrami opened the pull request:
Adding support for match ... case in polatify by converting match ... case into a nested ast.If.
<!-- ⚠️ This is an open-source repository. Do not share sensitive information. -->

pavelzw commented:
@prsabahrami could you please take care of the merge conflicts?

prsabahrami commented:
@pavelzw It should be fixed now.

pavelzw commented:
Thanks for your effort! A couple of notes:

---

The test failures are because of a new pixi version, #61 fixed them 😅 could you merge `main` again please?

---

`match ... case` is not 100% covered yet, for example this function doesn't work:

```py
def match_with_or(x):
    match x:
        case 0 | 1:
            return 0
        case 2:
            return 2 * x
        case 3:
            return 3 * x
    return x
```

---

Also, I personally would prefer match case being translated into consecutively chained `pl.when(...).then(...).when(...).then(...)...otherwise(...)` as this matches the original AST with `match ... case` better.

For example:

```py
def match_signum(x):
    s = 0
    match x:
        case 0:
            s = 1
        case 2:
            s = -1
        case 3:
            s = 0
    return s

# gets transformed right now to
def match_signum_polarified(x):
    import polars as pl
    return pl.when(x == 0).then(1).otherwise(pl.when(x == 2).then(-1).otherwise(pl.when(x == 3).then(0).otherwise(0)))

# imo this would be better
def match_signum_polarified(x):
    import polars as pl
    return pl.when(x == 0).then(1)\
        .when(x == 2).then(-1)\
        .when(x == 3).then(0)\
        .otherwise(0) # implicit otherwise case because no `case _` is given
```

---

It would also be nice to have a test case for the `match _` case.

---

Additionally we either need to implement some logic to differentiate between python versions and only include this test for python >=3.10 or just drop python 3.9 support altogether. Thoughts @0xbe7a?

0xbe7a commented:
I think we should still support py3.9

prsabahrami commented:
Thank you very much for the feedback!

I did merge the main and it seems like we still have some errors. 

Also, good idea to tranform it into consecutively chained `pl.when(...).then(...).when(...).then(...)...otherwise(...)` I'll try tp implement `match _` case as well as the consecutively chained translation today!

pavelzw commented:
> I did merge the main and it seems like we still have some errors.

This is because of python3.9 not supporting `match ... case`. You can ignore that for now and focus on implementing the missing features. Afterwards, I can assist you with adding py39 support.

The `py312` test should work, though: `pixi run -e py312 test`

prsabahrami commented:
I have added both functinalities both `match _` case and support for below example:

```py
def match_with_or(x):
    match x:
        case 0 | 1:
            return 0
        case 2:
            return 2 * x
        case 3:
            return 3 * x
    return x
```

**Note**:  I have changed the `Conditional State` s.t. it contains a list of "test" and a list of "then". This will affect `If`'s functionality as well. I could have created an additional `State`. I wasn't sure which one would be more suitable to your future needs and requirements but it can easily be updated. 
Also, the example given above will be translated to:
```py
def match_with_or_polarified(x):
    import polars as pl
    return pl.when((x == 0) |  (x == 1)).then(0).when(x == 2).then(2 * x).when(x == 3).then(3 * x).otherwise(x)
```
I believe the extra parentheses can be removed via unparsing and parsing and unparsing the result. However, let me know if you can think of any better solution!

0xbe7a commented:
Looks great, little nit-pick: Instead of `test: Sequence[ast.expr], then: Sequence[State]` we can use `Sequence[tuple[ast.Expr, State]]` instead, as this would provide a little more type safety.

pavelzw commented:
I found another non-covered case that throws a weird error message 😅 

```py
def match_a(x):
    y = 0
    match x, y:
        case 1, 0:
            return 4
        case _:
            return 5
```

```
E               AttributeError: 'Tuple' object has no attribute 'id'

polarify/main.py:198: AttributeError
```

prsabahrami commented:
I haven’t yet added the support for all `ast.MatchSequence()` or `ast.MatchMapping()`. I will try to add support for all cases and the change you suggested by today!

prsabahrami commented:
I have made the change from `test: Sequence[ast.expr], then: Sequence[State]` to `Sequence[tuple[ast.Expr, State]]`. Also, I have added support and tests for py 3.9.

**Few notes:** I have added support for below examples:
```py
def match_a(x):
    match x:
        case 1, 0:
            return 4
        case _:
            return 5
 ```
 and 
```py
def match_a(x):
    match x:
        case [1, 0]:
            return 4
        case _:
            return 5
 ```
 and 
 ```py
def match_a(x):
    match x:
        case [1, 0, *other]:
            return other
        case _:
            return 5
```

Regarding the above example: for now, it raises "starred patterns are not supported".
But to see its functionality comment the line 180 in `main.py`
```py
 180        raise ValueError("starred patterns are not supported")
```
and uncomment the two below lines
```py
 181        self.node.assignments[stmt.patterns[-1].name] = ast.Subscript(value=ast.Name(id=subj, ctx=ast.Load()), slice=ast.Slice(lower=ast.Constant(value=len(stmt.patterns) - 1)))
 182        left = ast.Subscript(value=ast.Name(id=subj, ctx=ast.Load()), slice=ast.Slice(upper=ast.Constant(value=len(stmt.patterns) - 1)))
```
Consider the below example:
```py
def match_a_polarified(x):
    import polars as pl
    return pl.when(x[:2] == [1, 0]).then(x[2:]).otherwise(5)
```
where this gives error since this is not how polars slices lists. I will be updating the above example to return something as below:
```py
def match_a_polarified(x):
    import polars as pl
    return pl.when(x.lower(2) == [1, 0]).then(x.slice(2, -1)).otherwise(5)
```
---

And I am adding the support for the below examples as well:
```py
def match_a(x):
    y = 0
    match x, y:
        case 1, 0:
            return 4
        case _:
            return 5
 ```
```py
def match_a(x):
    y = 3
    match x:
        case y if y > 5:
            return 1
        case _:
            return 5
 ```

prsabahrami commented:
I have added the below functionalities as well:
```py
def match_with_guard(x):
    match x:
        case y if y > 5:
            return 1
        case _:
            return 5
 ```
 will be translated to:
 ```py
 def match_with_guard_polarified(x):
    import polars as pl
    return pl.when(x > 5).then(1).otherwise(5)
```
However it still does not support lists yet and only supports single variable guard. (Warnings should be added!)
Also, it supports multi-variable match as below as well:
```py
def match_a(x):
    y = 0
    match x, y:
        case 1, 0:
            return 4
        case _:
            return 5
 ```
 will be translated to 
 ```py
 def match_a_polarified(x):
    import polars as pl
    return pl.when((x == 1) & (0 == 0)).then(4).otherwise(5)
```

Please let me know if you find any bugs or cases missing! 
Given I have added numerous cases, **there are warnings associated with each of these cases which should be added.**

pavelzw commented:
Sorry for the delays, i'll get back to you soon, i'm just a bit busy right now 😓

prsabahrami commented:
> Sorry for the delays, i'll get back to you soon, i'm just a bit busy right now 😓

No worries! I am working on fixing some issues but I'm a bit busy here as well so it might take two-three days!

pavelzw commented on polarify/main.py:
```diff
@@ -182,9 +310,12 @@ def parse_body(full_body: list[ast.stmt], assignments: dict[str, ast.expr] | Non
             state.handle_return(stmt.value)
             break
+        elif isinstance(stmt, ast.Match):
+            if PY_39:
+                raise ValueError("match statements are only supported in Python 3.9+")
```
```suggestion
            assert not PY_39
```

prsabahrami replied:
Hello,

Thanks a lot for the suggestions!
I am currently in the middle of my finals, I will look over the comments after after Apr. 23rd.

Thanks,

pavelzw reviewed (commented):
haven't reviewed everything yet, some comments

prsabahrami commented on polarify/main.py:
```diff
@@ -75,6 +84,25 @@ def visit_Compare(self, node: ast.Compare) -> ast.Compare:
+            node.lower = self.visit(node.lower)
+        if node.upper:
+            node.upper = self.visit(node.upper)
+        return node
```
I wanted to use them for when polars matches x with lists. for example:
```py
def match_x(x):
   match x:
       case [1, 2, 3]:
           return 1
       case [1, 4, *]:
           return 2
       case _:
           return 3
  ```
  However, I am raising an unsupported error right now. But, I plan to add these functionalities as well.
  *_The functions are deleted in the latest commit_*

prsabahrami commented:
Hello,

Was wondering if you could review the latest commit and approve?

Thanks

prsabahrami commented:
Increased the code coverage to ``0.9575``
Also, added some tests both for >= ``PY310`` and ``PY39``. 
Everything should be fine now.

prsabahrami commented:
Added the below test for coverage of L329 - L333:

```py

def multiple_match(x):
    match x:
        case 0:
            return 1
        case 1:
            return 2
    match x:
        case 0:
            return 3
        case 1:
            return 4
    return 5
```

0xbe7a commented on polarify/main.py:
```diff
@@ -139,25 +186,104 @@ class State:
+            return self.translate_match(subj, ast.MatchSequence(patterns=[stmt]))
+        elif isinstance(stmt, ast.MatchAs) and isinstance(subj, ast.Name):
+            if stmt.name is not None:
+                self.handle_assign(
```
Is there any valid code that requires this?

prsabahrami replied:
I was thinking of something similar to below:
```py
def match_x(x):
    match x:
        case _ if x > 5:
            return 0
        case _:
            return 1
```
In which case `ast` returns an empty `MatchAs` in which case the `if` statement on line 215 is required.

0xbe7a commented on polarify/main.py:
```diff
@@ -139,25 +186,104 @@ class State:
+                while len(subj.elts) > len(stmt.patterns):
+                    stmt.patterns.append(ast.MatchValue(value=ast.Constant(value=None)))
+                left = self.translate_match(subj.elts[0], stmt.patterns[0], guard)
+                right = (
```
This is not equivalent to the python behavior, see the test `match_sequence_padded_length`

0xbe7a replied:
Fine for now, we should use polars list operations in the future

0xbe7a reviewed (commented):
This mostly looks good, i found some failing test-cases

prsabahrami commented:
**Correct me if I'm wrong:** Based on my understanding python ignores cases where the number of subjects is not equal to the number of matching values. 
```py
def match_sequence_padded_length(x):
    y = 2
    z = None

    match x, y, z:
        case 1, 2:
            return 1
        case _:
            return x
 ```
 essentially returns x. 
 
 I have fixed the failing cases and for the below tests I am raising `ValueError` given that testing does not support literal returns:
 def match_sequence_padded_length_no_case(x):
    y = 2
    z = None

    match x, y, z:
        case 1, 2:
            return 1
        case _:
            return 2
 ```

prsabahrami commented:
I was wondering if you had the time to go over the changes? Any new suggestions or failing test cases?

0xbe7a commented:
No not yet, i hope we can take a look at it tomorrow @pavelzw

pavelzw reviewed (commented):
Looks very good, after the minor comments i think this is ready to merge 🥳

pavelzw reviewed (commented):
Looks very good, after the minor comments i think this is ready to merge 🥳

prsabahrami commented on polarify/main.py:
```diff
@@ -139,25 +193,101 @@ class State:
+        guard: ast.expr | None = None,
+    ):
+        """
+        TODO: Explain the purpose and goal of this method, it's quite complex
```
Done.

prsabahrami commented:
I added the explanation comment in translate_match. 
I believe we should be able to merge now.

pavelzw commented:
Thanks for enduring this @prsabahrami 😅 match statements in python are a lot more involved than I initially thought 😄

prsabahrami commented:
Of course!
It was such a great experience!
